services/src/contexts/migrations/database_migration.rs

```diff
@@ -0,0 +1,300 @@
+    async fn migrate(&self, tx: &Transaction<'_>) -> Result<()>;
+}
+
+/// The current version of the database. `None` if the database is empty.
+async fn determine_current_database_version<Tls>(
+    conn: &PooledConnection<'_, PostgresConnectionManager<Tls>>,
+) -> Result<Option<DatabaseVersion>>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    let stmt = match conn
+        .prepare("SELECT database_version from geoengine;")
+        .await
+    {
+        Ok(stmt) => stmt,
+        Err(e) => {
+            if let Some(code) = e.code() {
+                if *code == SqlState::UNDEFINED_TABLE {
+                    return Ok(None);
+                }
+            }
+            return Err(crate::error::Error::TokioPostgres { source: e });
+        }
+    };
+
+    let row = conn.query_one(&stmt, &[]).await?;
+
+    Ok(Some(row.get(0)))
+}
+
+#[derive(Debug, PartialEq)]
+pub enum MigrationResult {
+    CreatedDatabase,
+    MigratedDatabase,
+    AlreadyUpToDate,
+}
+
+/// Migrate the database to the latest version. If the database is empty, the initial migration is applied.
+///
+/// # Panics
+///
+/// Panics if there is an error in the migration logic.
+pub async fn migrate_database<Tls>(
+    conn: &mut PooledConnection<'_, PostgresConnectionManager<Tls>>,
+    migrations: &[Box<dyn Migration>],
+) -> Result<MigrationResult>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    let pre_migration_version = determine_current_database_version(conn).await?;
+    info!("Current database version: {:?}", pre_migration_version);
+
+    // start with the first migration after the current version
+    let applicable_migrations = migrations
+        .iter()
+        .skip_while(|m| m.prev_version() != pre_migration_version);
+
+    let mut current_version = pre_migration_version.clone();
+
+    for migration in applicable_migrations {
+        ensure!(
+            migration.prev_version() == current_version,
+            UnexpectedDatabaseVersionDuringMigration {
+                expected: migration.prev_version().unwrap_or_default(),
+                found: current_version.unwrap_or_default()
+            }
+        );
+
+        info!("Applying migration: {}", migration.version());
+
+        let tx = conn.build_transaction().start().await?;
+
+        migration.migrate(&tx).await?;
+
+        let stmt = tx
+            .prepare("UPDATE geoengine SET database_version = $1")
+            .await?;
+
+        tx.execute(&stmt, &[&migration.version()]).await?;
+
+        tx.commit().await?;
```

Comment:
Maybe, we could check at the end, that all workflows adhere to our JSON schema? We could determine faulty migrations this way.

Reply:
That's possible but could also be very costly. If you have thousands of workflows stored you have to load them all from the database and parse them.

I think it suffices to write tests for the migrations that actually change something in the operators and test exactly what they change to not break anything.

Reply:
1. The necessity of loading it out of the database means you ignored the task hint for pg_jsonschema (https://github.com/supabase/pg_jsonschema). This way, you don't need to load it out to check it.

2. Could you point to where you checked this?

Reply:
Johannes: write a todo and/or make new card

Reply:
I created a new card https://trello.com/c/SH9dewSd